cmd/kubeadm/app/apis/kubeadm/types.go

```diff
@@ -455,5 +455,34 @@ type ComponentConfig interface {
 	Get() interface{}
 }
 
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// ResetConfiguration contains a list of elements that is specific "kubeadm reset"-only runtime information.
+type ResetConfiguration struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// SkipPhases is a list of phases to skip during command execution.
+	// The list of phases can be obtained with the "kubeadm reset phase --help" command.
+	// The flag "--skip-phases" takes precedence over this field.
+	SkipPhases []string
+
+	// ForceReset flag instructs kubeadm to reset the node without prompting for confirmation
+	// The flag "--force" takes precedence over this field.
+	ForceReset bool
+
+	// CertificatesDir specifies where to store or look for all required certificates.
+	// Cannot mix the config for both the resetConfig and flag to make it consistent with kubeadm init.
+	CertificatesDir string
+
+	// CRISocket is used to retrieve container runtime info. This information will be annotated to the Node API object, for later re-use.
+	CRISocket string
```

Comment:
CriSocket can be specified in `init`/`join`/reset. 
For `upgrade`, it will use the node cri annotation. 
It will also be saved in the `InitConfiguration.nodeRegistration.criSocket`. 

Before `kubeadm reset`
- will take the `--cri-socket` if specified
- then it will take the `InitConfiguration.NodeRegistration.CRISocket` if it can get it
- the next is auto-detect

So I prefer to think that we can remove the `CRISocket` in reset configuration.
- If the kubeadm can retrieve `ResetConfiguration`, it can also get the `InitConfiguration`.

Reply:
i think we already have --cri-socket for reset? it makes sense to have the config field too.

we could obtain the cri socket from node annotation during reset, but one difference with upgrade is that upgrade must pass, while reset os best effort.

if reset cannot obtain the cri socket from thr node it will not be able to delete containers. and to obtain the node we need the node name. IIRC, currently this is done with a hack that looks at kubelet subject name in a certificate. note: user can override node name and it can be different from host name 

so if the cert is deleted for some reason reset cannot delete running containers as there is no socket to use.

this is just on the top of my head, and i could be missing details / be wrong.

Reply:
hm, actually i think we already construct InitConfiguration during reset by fetching node name, etc. i don't remember.

> i think we already have --cri-socket for reset? it makes sense to have the config field too.

this point still stands, i guess.

Reply:
i'd swap 2 and 3, since 3 is explicit config for reset.

Reply:
> 3\. it will check `ResetConfiguration.CRISocket` from configmap.

hm, but we never discussed storing ResetConfiguration in a CM.

Reply:
> hm, but we never discussed storing ResetConfiguration in a CM.

Got it now. 
- flag > config inline > cluster configuration(from CM) > auto-detect.

Reply:
> so if the cert is deleted for some reason reset cannot delete running containers as there is no socket to use.

It can fall back to auto-detect on host as well, but I am still in favor of the configure of criScoket in the config file, 

> i think we already have --cri-socket for reset? it makes sense to have the config field too.

agreed!

> 1. it will take the --cri-socket if specified
> 2. it will take the InitConfiguration.NodeRegistration.CRISocket from configmap or node annotation.
> 3. it will check ResetConfiguration.CRISocket from configmap.
> 4. the next is auto-detect.

I think 2 is not supported in this patch, the sequential now is

1. is `--cri-socket` if specified
2. is  `ResetConfiguration.CRISocket` configured?
3. let's auto-detect on this host.

Do we still need check socket from the initCfg?  pls let me know.

Reply:
> I think 2 is not supported in this patch

i think it was, because reset fetches the clusterconfiguration from kubeadm-config CM, then it constructs an initconfiguration from what information it has about the node. this same information should contain the CRI socket.

Reply:
@neolit123 

sorry, I express it not clear enough, what I meant is that we haven't check the CRI socket from the  `InitConfiguration.NodeRegistration.CRISocket` in this patch.


https://github.com/kubernetes/kubernetes/pull/113583/files#diff-d797a5d9b766ed766209920af2aff80b4a8cf503d6109710ac9b3a8c2b6dca0fR136-R143


Should us get the CRI socket from `InitConfiguration.NodeRegistration.CRISocket` before the auto-detect on host? If yes, I will add the logic in this pr. Seems like to me a little redundant to do that but I am fine to add that logic.

Reply:
do we not fetch ClusterConfiguration from kubeadm-config cm during reset?

i think we do. if yes, then that must also construct an InitConfiguration

Reply:
Done